File: sdks/ts/memory/src/ingest/hash.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * BLAKE3 and SHA-256 hashing utilities for the ingest pipeline.
 * BLAKE3 is used for all new documents. SHA-256 is retained for
 * dual-read compatibility during the migration period.
 *
 * Provides deterministic 256-bit BLAKE3 hashes hex-encoded to 64-character
 * strings. Used for document deduplication and chunk-level change detection.
 * Cross-SDK conformant: identical inputs produce identical outputs in both
 * the Go and TypeScript SDKs.
 */

import { createHash } from 'node:crypto'
import { blake3 } from '@noble/hashes/blake3.js'
import { bytesToHex } from '@noble/hashes/utils.js'

/**
 * Compute the BLAKE3 hash of a buffer. Returns the full hex digest
 * (64 characters).
 */
export const hashContent = (buf: Buffer): string => bytesToHex(blake3(buf))

/**
 * Compute the SHA-256 hash of a buffer. Returns the full hex digest.
 * Used for dual-read fallback during the BLAKE3 migration period.
 */
export const hashContentSHA256 = (buf: Buffer): string =>
  createHash('sha256').update(buf).digest('hex')

/**
 * Compute a truncated SHA-256 hash (12 hex chars). Matches the legacy
 * hashSlug used prior to the BLAKE3 migration.
 */
export const hashSlugSHA256 = (data: Buffer): string => hashContentSHA256(data).slice(0, 12)

/**
 * Hasher abstracts a content-hashing algorithm so callers can swap
 * implementations (BLAKE3, SHA-256, etc.) without changing call sites.
 */
export type Hasher = {
  /** Compute a hex-encoded digest of content. */
  hash(content: Buffer): string
  /** Short identifier for the algorithm (e.g. "blake3"). */
  name: string
}

/**
 * Default BLAKE3 Hasher implementation. 256-bit output, hex-encoded to
 * 64 lowercase characters.
 */
export const blake3Hasher: Hasher = {
  hash(content: Buffer): string {
    const digest = blake3(content)
    return bytesToHex(digest)
  },
  name: 'blake3',
}

/**
 * Compute the BLAKE3 256-bit hash of document content.
 * Returns a 64-character lowercase hex string.
 *
 * Time: O(n) where n = content.length.
 * Space: O(1) beyond the 32-byte digest.
 */
export const hashDocument = (content: Buffer): string => {
  const digest = blake3(content)
  return bytesToHex(digest)
}

/**
 * Compute the BLAKE3 256-bit hash of chunk content.
 * Semantically identical to hashDocument but named separately for clarity
 * at call sites performing chunk-level change detection.
 *
 * Time: O(n) where n = content.length.
 * Space: O(1) beyond the 32-byte digest.
 */
export const hashChunk = (content: Buffer): string => {
  const digest = blake3(content)
  return bytesToHex(digest)
}

/**
 * Compute the BLAKE3 256-bit hash of a string input.
 * Convenience wrapper that encodes the string as UTF-8 before hashing.
 * Returns a 64-character lowercase hex string.
 *
 * Time: O(n) where n = byte length of the UTF-8 encoding.
 * Space: O(n) for the intermediate buffer.
 */
export const hashString = (s: string): string => {
  const buf = Buffer.from(s, 'utf8')
  const digest = blake3(buf)
  return bytesToHex(digest)
}

/**
 * Compute a truncated BLAKE3 hash (12 hex chars) suitable for use as
 * a slug fallback. Matches the Go SDK's HashSlug function.
 *
 * Time: O(n) where n = data.length.
 * Space: O(1) beyond the 32-byte digest.
 */
export const hashSlug = (data: Buffer): string => hashDocument(data).slice(0, 12)

/**
 * Compute a stable document identifier from brain ID and content hash.
 * Returns 16 hex characters of BLAKE3(brainId + ":" + contentHash).
 * Matches the Go SDK's HashDocumentID function.
 *
 * Time: O(n) where n = brainId.length + contentHash.length.
 * Space: O(n) for the intermediate buffer.
 */
export const hashDocumentId = (brainId: string, contentHash: string): string => {
  const input = Buffer.from(`${brainId}:${contentHash}`, 'utf8')
  return hashDocument(input).slice(0, 16)
}

File: go/ingest/hash.go
// SPDX-License-Identifier: Apache-2.0
package ingest

import (
	"crypto/sha256"
	"encoding/hex"

	"github.com/zeebo/blake3"
	lukechampineblake3 "lukechampine.com/blake3"
)

// HashContent computes the BLAKE3 hash of data and returns the full
// hex-encoded digest (64 characters).
func HashContent(data []byte) string {
	sum := lukechampineblake3.Sum256(data)
	return hex.EncodeToString(sum[:])
}

// HashContentSHA256 computes the SHA-256 hash of data and returns the
// full hex-encoded digest. Used during migration to look up documents
// stored under the legacy SHA-256 scheme.
func HashContentSHA256(data []byte) string {
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

// HashSlugSHA256 computes the SHA-256 hash of data and returns the
// first 12 hex characters. Matches the legacy hashSlug function from
// knowledge/ingest.go.
func HashSlugSHA256(data []byte) string {
	return HashContentSHA256(data)[:12]
}

// Hasher abstracts a content-hashing algorithm so callers can swap
// implementations (BLAKE3, SHA-256, etc.) without changing call sites.
type Hasher interface {
	// Hash computes a hex-encoded digest of content.
	Hash(content []byte) string
	// Name returns a short identifier for the algorithm (e.g. "blake3").
	Name() string
}

// BLAKE3Hasher implements [Hasher] using the BLAKE3 algorithm with a
// 256-bit output, hex-encoded to 64 lowercase characters.
type BLAKE3Hasher struct{}

// Hash implements [Hasher].
func (BLAKE3Hasher) Hash(content []byte) string {
	sum := blake3.Sum256(content)
	return hex.EncodeToString(sum[:])
}

// Name implements [Hasher].
func (BLAKE3Hasher) Name() string { return "blake3" }

// DefaultHasher is the package-level Hasher used by all public hash
// functions. Defaults to BLAKE3.
var DefaultHasher Hasher = BLAKE3Hasher{}

// HashDocument computes the BLAKE3 hash of document content and returns
// the full hex-encoded 256-bit digest (64 characters). Used for document
// deduplication in the ingest pipeline.
func HashDocument(content []byte) string {
	sum := blake3.Sum256(content)
	return hex.EncodeToString(sum[:])
}

// HashChunk computes the BLAKE3 hash of chunk content and returns the
// full hex-encoded 256-bit digest (64 characters). Semantically identical
// to HashDocument but named separately for clarity at call sites where
// chunk-level change detection is the intent.
func HashChunk(content []byte) string {
	return HashDocument(content)
}

// HashString computes the BLAKE3 hash of a string input and returns the
// full hex-encoded 256-bit digest (64 characters). Convenience wrapper
// that avoids a []byte conversion at call sites.
func HashString(s string) string {
	sum := blake3.Sum256([]byte(s))
	return hex.EncodeToString(sum[:])
}

// HashSlug computes the BLAKE3 hash of data and returns the first 12
// hex characters. Used as a deterministic fallback slug when a
// human-readable slug is unavailable.
func HashSlug(data []byte) string {
	return HashDocument(data)[:12]
}

// HashDocumentID computes a stable, brain-scoped document identifier
// from a brain ID and a content hash. Returns the first 16 hex
// characters of BLAKE3(brainID + ":" + contentHash).
func HashDocumentID(brainID, contentHash string) string {
	input := []byte(brainID + ":" + contentHash)
	return HashDocument(input)[:16]
}
